### ISSUE TITLE
Inner forest polygon in forest multipolygon not rendered

### ISSUE THREAD

**Issue by Penegal** [gravitystorm/openstreetmap-carto#2381]
Hello, there.

I noticed what seems to be a rendering bug in forest multipolygons: I modeled [here](https://www.openstreetmap.org/#map=18/48.24100/6.17774) a [needle-leaved forest parcel](https://www.openstreetmap.org/way/444365950) in a bigger broad-leaved forest as an inner multipolygon, but this inner parcel is not rendered, as if there was nothing here, a big hole, full of nothing, in the middle of the forest. I couldn't confirm this is a rendering problem, but I assume so, as jOSM and OSMose don't see any problem there, and iD correctly displays this area in edition.

Regards.

**Comment by jojo4u** [comment]
JOSM reports no problems.
Either some special problem because relation and inner has landuse=forest or a diff update problem. Try wiggeling some nodes and re-upload.

**Comment by math1985** [comment]
@pnorman Could this be the osm2pgsql multipolygon handling?

**Comment by SomeoneElseOSM** [comment]
For info see also https://help.openstreetmap.org/questions/52242/needleleaved-inner-forest-polygon-not-rendered-rendering-or-modelling-problem (in case anyone hasn't seen it).

**Comment by pnorman** [comment]
> @pnorman Could this be the osm2pgsql multipolygon handling?

It's possible, and it's also possible if there is a bug that it was fixed in more recent versions of osm2pgsql.

Either way, not a style bug that can be fixed in this repository.

**Comment by math1985** [comment]
> Either way, not a style bug that can be fixed in this repository.

I'm not sure if I agree, if it's an osm2pgsql multigolygon bug we might be able to fix it in the lua transformation file.

**Comment by imagico** [comment]
For info:

http://mc.bbbike.org/mc/?lon=6.177117&lat=48.23868&zoom=15&num=3&mt0=mapnik-german&mt1=mapnik&mt2=osmfr

It's in the french style as well but not in the german style.

**Comment by pnorman** [comment]
> I'm not sure if I agree, if it's an osm2pgsql multigolygon bug we might be able to fix it in the lua transformation file.

A better way to put it is that the bug isn't in osm-carto right now, but the Lua transforms have the potential to introduce bugs into our code which involve multipolygons. The multipolygon handling code of our transforms will certainly need testing, but this is one of those issues I suspect is long-fixed. So, without a specific bug report, I think **this** issue is closed.

**Comment by SomeoneElseOSM** [comment]
@pnorman I've no idea what "but the Lua transforms have the potential to introduce bugs into our code" is supposed to mean.  It's certainly true as @math1985 says that it would be possible "fix it in the lua transformation file" (at least this particular broadleaved/needleleaved instance - though more difficult to do generically) as I don't see the issue in an osm-carto derived style and a recent osm2pgsql with a lua transform that treats broadleaved and needleleaved as different tags..

Where do you suggest that the issue should be logged (because there clearly is one - I can see the hole just as anyone else can)?

**Comment by pnorman** [comment]
> @pnorman I've no idea what "but the Lua transforms have the potential to introduce bugs into our code" is supposed to mean

Right now osm-carto has no control over multipolygon processing. It is all done by osm2pgsql. This means we **can't** have multipolygon bugs here[1]. The proposed lua changes will mean that we can have MP bugs, but they're not the cause of this one because they're not merged yet.

> Where do you suggest that the issue should be logged (because there clearly is one - I can see the hole just as anyone else can)?

If it can be reproduced with the latest osm2pgsql release it should be reported on the osm2pgsql issue tracker. Given the German style doesn't show the issue I expect its fixed.

[1] Unless we started doing crazy stuff like manipulating geometries

**Comment by SomeoneElseOSM** [comment]
Do we know what osm2pgsql versions are live on osm.org and on the German tile servers currently?

**Comment by gravitystorm** [comment]
This is almost certainly based on osm2pgsql's handling of "old-style" multipolygon tagging. It used to be possible to tag multipolygons on the outer ring, with no tags on the relation. If the same tags are found on the inner ring, these are assumed to be duplicates of the outer ring's tags. Given that we aren't using the leaf_* tags in our osm2pgsql.style file, I bet osm2pgsql is assuming this is just a duplicate tagging of the outer way, and ignoring its potential to be a second, nested, polygon.

I wouldn't be surprised to find that with different contents of osm2pgsql.style files (i.e. including the leaf_* tags) or with hstore enabled, osm2pgsql would create the second polygon. That could explain differences between tileservers. Or it could be different versions of osm2pgsql handling things differently too.

In short, there are so many moving parts here, it's probably better to see if it works as expected on the lua branch, than to try to figure out why it's happening with our current setup.

**Comment by d1g** [comment]
@gravitystorm, Andy please copy&paste this info somewhere visible (README.md/LIMITATIONS.md) - otherwise we will repeat questions: #2297

**Comment by Penegal** [comment]
Looks like PR #2533 solves the problem: https://lua.osm-carto.paulnorman.ca/#15/48.2408/6.1777
